Mathlib/RingTheory/FractionalIdeal/Norm.lean

```diff
@@ -0,0 +1,145 @@
+      ← abs_mul, Int.cast_mul]
+
+theorem absNorm_eq (I : FractionalIdeal R⁰ K) :
+    absNorm I = (Ideal.absNorm I.num : ℚ) / |Algebra.norm ℤ (I.den : R)| := rfl
+
+theorem absNorm_eq' {I : FractionalIdeal R⁰ K} (a : R⁰) (I₀ : Ideal R)
+    (h : a • (I : Submodule R K) = Submodule.map (Algebra.linearMap R K) I₀) :
+    absNorm I = (Ideal.absNorm I₀ : ℚ) / |Algebra.norm ℤ (a:R)| := by
+  rw [absNorm, ← absNorm_div_norm_eq_absNorm_div_norm a I₀ h, MonoidWithZeroHom.coe_mk,
+    ZeroHom.coe_mk]
+
+theorem coeIdeal_absNorm (I₀ : Ideal R) :
+    absNorm (I₀ : FractionalIdeal R⁰ K) = Ideal.absNorm I₀ := by
+  rw [absNorm_eq' 1 I₀ (by rw [one_smul]; rfl), OneMemClass.coe_one, _root_.map_one, abs_one,
+    Int.cast_one, _root_.div_one]
+
+theorem absNorm_bot : absNorm (⊥ : FractionalIdeal R⁰ K) = 0 := absNorm.map_zero'
+
+theorem absNorm_eq_zero_iff [NoZeroDivisors K] {I : FractionalIdeal R⁰ K} :
+    absNorm I = 0 ↔ I = 0 := by
+  refine ⟨fun h ↦ zero_of_num_eq_bot zero_not_mem_nonZeroDivisors ?_, fun h ↦ h ▸ absNorm_bot⟩
+  rw [absNorm_eq, div_eq_zero_iff] at h
+  refine Ideal.absNorm_eq_zero_iff.mp <| Nat.cast_eq_zero.mp <| h.resolve_right ?_
+  simpa [Algebra.norm_eq_zero_iff] using nonZeroDivisors.coe_ne_zero _
+
+section IsLocalization
+
+variable [IsLocalization (Algebra.algebraMapSubmonoid R ℤ⁰) K] [Algebra ℚ K]
+
+theorem abs_det_basis_change [NoZeroDivisors K] {ι : Type*} [Fintype ι] [DecidableEq ι]
+    (b : Basis ι ℤ R) (I : FractionalIdeal R⁰ K) (bI : Basis ι ℤ I) :
+    |(b.localizationLocalization ℚ ℤ⁰ K).det ((↑) ∘ bI)| = absNorm I := by
+  let b₀ : Basis ι ℚ K := b.localizationLocalization ℚ ℤ⁰ K
+  let bI.num : Basis ι ℤ I.num := bI.map
+      ((equivNum (IsFractionRing.injective R K) (nonZeroDivisors.coe_ne_zero _)).restrictScalars ℤ)
+  rw [absNorm_eq, ← Ideal.natAbs_det_basis_change b I.num bI.num, Int.cast_natAbs, Int.cast_abs,
+    Int.cast_abs, Basis.det_apply, Basis.det_apply]
+  change _ = |algebraMap ℤ ℚ _| / _
+  rw [RingHom.map_det, show RingHom.mapMatrix (algebraMap ℤ ℚ) (b.toMatrix ((↑) ∘ bI.num)) =
+      b₀.toMatrix ((algebraMap R K (den I : R)) • ((↑) ∘ bI)) by
+    ext : 2
+    simp_rw [RingHom.mapMatrix_apply, Matrix.map_apply, Basis.toMatrix_apply,
+      ← Basis.localizationLocalization_repr_algebraMap ℚ ℤ⁰ K, Function.comp_apply,
+      Basis.map_apply, LinearEquiv.restrictScalars_apply, equivNum_apply, Submonoid.smul_def,
+      Algebra.smul_def]
+    rfl]
+  rw [Basis.toMatrix_smul, Matrix.det_mul, abs_mul, ← Algebra.norm_eq_matrix_det,
+    Algebra.norm_localization ℤ ℤ⁰, show (Algebra.norm ℤ (den I: R) : ℚ) =
+    algebraMap ℤ ℚ (Algebra.norm ℤ (den I: R)) by rfl, mul_div_assoc, mul_div_cancel' _ (by
+    rw [ne_eq, abs_eq_zero, IsFractionRing.to_map_eq_zero_iff, Algebra.norm_eq_zero_iff_of_basis b]
+    exact nonZeroDivisors.coe_ne_zero _)]
+
+@[simp]
+theorem absNorm_span_singleton [IsDomain K] [Module.Finite ℚ K] (x : K) :
```

Comment:
I am a bit confused by `IsDomain K`. Wasn't `K` the fraction field of `R`?

Reply:
Yes. thanks. This is unnecessary